From bce37e345b555c38cf2adc57b13b62ad5532df50 Mon Sep 17 00:00:00 2001
From: John Benediktsson <mrjbq7@gmail.com>
Date: Thu, 18 Sep 2025 12:18:42 -0700
Subject: [PATCH 105/133] std.c: adjust shm_open to be variadic on darwin
X-Developer-Signature: v=1; a=openpgp-sha256; l=2000; i=xtex@astrafall.org;
 h=from:subject; bh=OMASPFKeG2GfvJXMozkPFn+i7ebnjEmgWr8GoNzPiTI=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78rjF0lmDE2+zt6nHB3IvaPZesLCs0ep9O2/Nq2vqR
 yVf1c7smMjCIMbFYCmmyFJk2ODNqpPOL7qsXBZmDisTyBBpkQYGBgYGFga+3MS8UiMdIz1TbUM9
 Q0MdIx0jBi5OAZhq7dUM/90Cd81liNJaZv3+26Oda5jWLtzRuMo35HX6oaL35pcVZVwZ/kqo3D5
 c4lz6JPCy4txnrL/OLA5fzWf0nPefQoDh3/XxMrwA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit 1ac4c27d74c95e31477a03557ad3807fd2bed57d)
---
 lib/std/c.zig        | 6 +++++-
 lib/std/c/darwin.zig | 1 +
 2 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/lib/std/c.zig b/lib/std/c.zig
index a3f98c75640c..03584f407333 100644
--- a/lib/std/c.zig
+++ b/lib/std/c.zig
@@ -10813,7 +10813,10 @@ pub extern "c" fn sem_trywait(sem: *sem_t) c_int;
 pub extern "c" fn sem_timedwait(sem: *sem_t, abs_timeout: *const timespec) c_int;
 pub extern "c" fn sem_getvalue(sem: *sem_t, sval: *c_int) c_int;
 
-pub extern "c" fn shm_open(name: [*:0]const u8, flag: c_int, mode: mode_t) c_int;
+pub const shm_open = switch (native_os) {
+    .driverkit, .macos, .ios, .tvos, .watchos, .visionos => darwin.shm_open,
+    else => private.shm_open,
+};
 pub extern "c" fn shm_unlink(name: [*:0]const u8) c_int;
 
 pub extern "c" fn kqueue() c_int;
@@ -11362,6 +11365,7 @@ const private = struct {
     extern "c" fn stat(noalias path: [*:0]const u8, noalias buf: *Stat) c_int;
     extern "c" fn sigaltstack(ss: ?*stack_t, old_ss: ?*stack_t) c_int;
     extern "c" fn sysconf(sc: c_int) c_long;
+    extern "c" fn shm_open(name: [*:0]const u8, flag: c_int, mode: mode_t) c_int;
 
     extern "c" fn pthread_setname_np(thread: pthread_t, name: [*:0]const u8) c_int;
     extern "c" fn getcontext(ucp: *ucontext_t) c_int;
diff --git a/lib/std/c/darwin.zig b/lib/std/c/darwin.zig
index cc94964a5f01..fc18210ab612 100644
--- a/lib/std/c/darwin.zig
+++ b/lib/std/c/darwin.zig
@@ -305,6 +305,7 @@ pub const mach_header = std.macho.mach_header;
 pub extern "c" fn @"close$NOCANCEL"(fd: fd_t) c_int;
 pub extern "c" fn mach_host_self() mach_port_t;
 pub extern "c" fn clock_get_time(clock_serv: clock_serv_t, cur_time: *mach_timespec_t) kern_return_t;
+pub extern "c" fn shm_open(name: [*:0]const u8, flag: c_int, ...) c_int;
 
 pub const exception_data_type_t = integer_t;
 pub const exception_data_t = ?*mach_exception_data_type_t;
-- 
2.52.0

